gh-130472: Integrate fancycompleter with the new repl, to get colored tab completions

antocuni opened the pull request:
As the title says, this is (WIP) to integrate fancycompleter into the new REPL.

<!-- gh-issue-number: gh-130472 -->
* Issue: gh-130472
<!-- /gh-issue-number -->

antocuni committed:
import fancycompleter from https://github.com/pdbpp/fancycompleter/commit/67e3ec128cf8d44be6e48e775234c07f4b23064e

antocuni committed:
add copyright notice

antocuni committed:
enable FancyCompleter by default, unless you set PYTHON_BASIC_COMPLETER

antocuni committed:
WIP: kill a lot of code which is no longer necessary

antocuni committed:
force colors for now

antocuni committed:
kill the logic to find a readline, we can always use _pyrepl.readline now

antocuni committed:
kill LazyVersion

antocuni committed:
we surely don't need to support python 2.7 now :)

antocuni committed:
kill ConfigurableClass

antocuni committed:
better name

antocuni committed:
use _colorize instead of our own Color

antocuni committed:
WIP: copy&adapt some tests from the original fancycompleter. They don't work because they need to be ported from pytest to unittest

antocuni committed:
edited by copilot: move from pytest-style to unittest-style

antocuni committed:
don't try to be too clever with exceptions: if a global name raises an exception when evaluated, just color it as None

antocuni committed:
no longer needed

antocuni committed:
this doesn't test anything meaningful

antocuni committed:
fix this test

antocuni committed:
fix this test

antocuni committed:
Fix this test

hugovk commented on Lib/_colorize.py:
```diff
@@ -7,15 +7,19 @@
     BOLD_GREEN = "\x1b[1;32m"
     BOLD_MAGENTA = "\x1b[1;35m"
     BOLD_RED = "\x1b[1;31m"
+    BOLD_TEAL = "\x1b[1;36m"
```
ANSI colour with code 36 is more commonly called cyan: https://en.wikipedia.org/wiki/ANSI_escape_code#Colors


```suggestion
    BOLD_BLUE = "\x1b[1;34m"
    BOLD_CYAN = "\x1b[1;36m"
    BOLD_GREEN = "\x1b[1;32m"
    BOLD_MAGENTA = "\x1b[1;35m"
    BOLD_RED = "\x1b[1;31m"
```

hugovk commented on Lib/_colorize.py:
```diff
@@ -7,15 +7,19 @@
     MAGENTA = "\x1b[35m"
     RED = "\x1b[31m"
     RESET = "\x1b[0m"
+    TEAL = "\x1b[36m"
```
```suggestion
    BLACK = "\x1b[30m"
    CYAN = "\x1b[36m"
    GREEN = "\x1b[32m"
    GREY = "\x1b[90m"
    MAGENTA = "\x1b[35m"
    RED = "\x1b[31m"
    RESET = "\x1b[0m"
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        types.MethodType: ANSIColors.BOLD_TEAL,
+        type((42).__add__): ANSIColors.BOLD_TEAL,
+        type(int.__add__): ANSIColors.BOLD_TEAL,
+        type(str.replace): ANSIColors.BOLD_TEAL,
```
```suggestion
        types.BuiltinMethodType: ANSIColors.BOLD_CYAN,
        types.MethodType: ANSIColors.BOLD_CYAN,
        type((42).__add__): ANSIColors.BOLD_CYAN,
        type(int.__add__): ANSIColors.BOLD_CYAN,
        type(str.replace): ANSIColors.BOLD_CYAN,
```

AA-Turner replied:
Hugo's suggestion & using types from `types`
```suggestion
        types.BuiltinMethodType: ANSIColors.BOLD_CYAN,
        types.MethodType: ANSIColors.BOLD_CYAN,
        types.MethodWrapperType: ANSIColors.BOLD_CYAN,
        types.WrapperDescriptorType: ANSIColors.BOLD_CYAN,
        types.MethodDescriptorType: ANSIColors.BOLD_CYAN,
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+
+        type: ANSIColors.BOLD_MAGENTA,
+
+        types.ModuleType: ANSIColors.TEAL,
```
```suggestion
        types.ModuleType: ANSIColors.CYAN,
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+#
+#                        All Rights Reserved
+"""
+Colorful TAB completion for Python prompt
```
```suggestion
Colorful tab completion for Python prompt
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+
+class Completer(rlcompleter.Completer):
+    """
+    When doing someting like a.b.<TAB>, display only the attributes of
```
```suggestion
    When doing someting like a.b.<tab>, display only the attributes of
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+    def complete(self, text, state):
+        """
+        stolen from:
+        http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/496812
```
This link doesn't work. Is there a live version? And a better word than "stolen"? :)

antocuni replied:
I tried to find it on the wayback machine but it seems to be down ATM.
Anyway we can either:
  - put a link to the wayback machine (assuming it exists)
  - remove the comment altogether

I'm happy with both.

hugovk commented on Lib/test/test_pyrepl/test_fancycompleter.py:
```diff
@@ -0,0 +1,192 @@
+        matches = compl.attr_matches('A.a')
+        self.assertEqual(sorted(matches), [' ', 'aaa', 'abc_1', 'abc_2', 'abc_3'])
+        #
+        # IF there is an actual common prefix, we return just it, so that readline
```
```suggestion
        # If there is an actual common prefix, we return just it, so that readline
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+            # this is needed to offer pyrepl a better chance to patch
+            # raw_input. Usually, it does at import time, but is we are under
+            # pytest with output captured, at import time we don't have a
+            # terminal and thus the raw_input hook is not installed
```
```suggestion
            # This is needed to offer PyREPL a better chance to patch
            # raw_input. Usually, it does at import time, but if we are under
            # pytest with output captured, at import time we don't have a
            # terminal and thus the raw_input hook is not installed.
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+
+    def _callable_postfix(self, val, word):
+        # disable automatic insertion of '(' for global callables:
+        # this method exists only in Python 2.6+
```
Doesn't matter in CPython 3.14 in 2025 :)
```suggestion
        # disable automatic insertion of '(' for global callables
```

antocuni replied:
if anything, this is a proof that fancycompleter has been battle tested for like ~18 years by now :)

hugovk commented on Lib/test/test_pyrepl/test_fancycompleter.py:
```diff
@@ -0,0 +1,192 @@
+        # in this case, we want to display all attributes which start with
+        # 'a'. MOREOVER, we also include a space to prevent readline to
+        # automatically insert the common prefix (which will the the ANSI escape
+        # sequence if we use colors)
```
```suggestion
        # In this case, we want to display all attributes which start with
        # 'a'. Moreover, we also include a space to prevent readline to
        # automatically insert the common prefix (which will the the ANSI escape
        # sequence if we use colors).
```

hugovk commented on Lib/test/test_pyrepl/test_fancycompleter.py:
```diff
@@ -0,0 +1,192 @@
+        #
+        # finally, at the next TAB, we display again all the completions available
+        # for this common prefix. Agai, we insert a spurious space to prevent the
+        # automatic completion of ANSI sequences
```
```suggestion
        # Finally, at the next tab, we display again all the completions available
        # for this common prefix. Again, we insert a spurious space to prevent the
        # automatic completion of ANSI sequences.
```

hugovk commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+                   for i, name, obj
+                   in zip(count(), names, values)]
+        # We add a space at the end to prevent the automatic completion of the
+        # common prefix, which is the ANSI ESCAPE sequence.
```
```suggestion
        # common prefix, which is the ANSI escape sequence.
```

hugovk reviewed (commented):
Exciting! Just some nits.

tomasr8 commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        # XXX: double check what happens in this case once fancycompleter works
+        if False and hasattr(readline, '_setup'):
+            # this is needed to offer pyrepl a better chance to patch
+            # raw_input. Usually, it does at import time, but is we are under
```
```suggestion
            # raw_input. Usually, it does at import time, but if we are under
```

tomasr8 commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+            return []
+
+        if len(names) == 1:
+            return ['%s.%s' % (expr, names[0])]  # only option, no coloring.
```
```suggestion
            return [f'{expr}.{names[0]}']  # only option, no coloring.
```

tomasr8 commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+
+        prefix = commonprefix(names)
+        if prefix and prefix != attr:
+            return ['%s.%s' % (expr, prefix)]  # autocomplete prefix
```
```suggestion
            return [f'{expr}.{prefix}']  # autocomplete prefix
```

tomasr8 commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+    def colorize_matches(self, names, values):
+        matches = [self.color_for_obj(i, name, obj)
+                   for i, name, obj
+                   in zip(count(), names, values)]
```
Maybe? This also lets you remove the `itertools` import.
```suggestion
                   for i, (name, obj)
                   in enumerate(zip(names, values))]
```

tomasr8 commented on Lib/_pyrepl/readline.py:
```diff
@@ -587,7 +588,12 @@ def _setup(namespace: Mapping[str, Any]) -> None:
         namespace = dict(namespace)
-    _wrapper.config.readline_completer = RLCompleter(namespace).complete
+
+    if os.getenv('PYTHON_BASIC_COMPLETER'):
```
Should this variable be documented?

antocuni replied:
yes, totally! I think it's worth discussing what is the best way to disable fancycompleter, if anybody has other ideas I'm happy to hear

antocuni replied:
documentation added in b710bce38de

tomasr8 commented on Lib/test/test_pyrepl/test_fancycompleter.py:
```diff
@@ -0,0 +1,192 @@
+    def test_unicode_in___dir__(self):
+        class Foo(object):
+            def __dir__(self):
+                return [u'hello', 'world']
```
```suggestion
                return ['hello', 'world']
```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+#   Copyright 2010-2025 Antonio Cuni
+#                       Daniel Hahler
+#
+#                        All Rights Reserved
```
Should this have a licence grant (e.g. [BSD-3-Clause](https://github.com/pdbpp/fancycompleter/blob/master/LICENSE))? If so, we should add it here and to the included software licences section.

AA-Turner replied:
Another option, which might be slightly nicer, is for the contributors of the relevant bits of `fancycompleter` to sign the CLA and have this work relicenced under the PSF licence. This seems to include @blueyed, @theY4Kman, @singingwolfboy, and @slinkp.

theY4Kman replied:
I'm happy to do so

slinkp replied:
I only made [this one obsolete commit to fancycompleter 13 years ago](https://github.com/pdbpp/fancycompleter/commit/90f9ca09e48bed3f1e3097b053a45f6f13c8144c), but i'll sign if it helps :)  How do I do that?

antocuni replied:
I'm not expert in licensing issue. I'm happy to do/sign whatever is needed to land the PR

StanFromIreland replied:
The CLA can be signed here: https://cla.python.org/

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+#                        All Rights Reserved
+"""
+Colorful TAB completion for Python prompt
+"""
```
```suggestion
"""Colorful tab completion for the Python prompt"""

```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+from _colorize import ANSIColors
+import rlcompleter
+import types
+from itertools import count
```
Can be removed with the `enumerate` suggestion
```suggestion
```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        type: ANSIColors.BOLD_MAGENTA,
+
+        types.ModuleType: ANSIColors.TEAL,
+        type(None): ANSIColors.GREY,
```
```suggestion
        types.NoneType: ANSIColors.GREY,
```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    def setup(self):
+        import _colorize
```
You already import `_colorize` at module level, no point in a lazy import

antocuni replied:
fixed by e2294ec8eab

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        self.config.setup()
+
+        # XXX: double check what happens in this case once fancycompleter works
+        if False and hasattr(readline, '_setup'):
```
We should fix this or remove the `if False` block

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/496812
+        """
+        if text == "":
+            return ['\t', None][state]
```
```suggestion
            return ('\t', None)[state]
```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        return word
+
+    def global_matches(self, text):
+        import keyword
```
Is a lazy import worth it here?

antocuni replied:
fixed by 70893239cf3

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+
+        # get the content of the object, except __builtins__
+        words = set(dir(thisobject))
+        words.discard("__builtins__")
```
```suggestion
        words = set(dir(thisobject)) - {'__builtins__'}
```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+        while True:
+            for word in words:
+                if (word[:n] == attr and
+                        not (noprefix and word[:n+1] == noprefix)):
```
Style:
```suggestion
                if (
                    word[:n] == attr 
                    and not (noprefix and word[:n+1] == noprefix)
                ):
```

AA-Turner commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,205 @@
+
+def commonprefix(names, base=''):
+    """ return the common prefix of all 'names' starting with 'base'
+    """
```
```suggestion
    """Return the common prefix of all 'names' starting with 'base'"""
```

AA-Turner commented on Lib/_pyrepl/readline.py:
```diff
@@ -35,6 +35,7 @@
 from site import gethistoryfile   # type: ignore[attr-defined]
 import sys
 from rlcompleter import Completer as RLCompleter
+from .fancycompleter import Completer as FancyCompleter
```
Sort this import below `from .console`?

antocuni replied:
fixed by 926c1a3611f

antocuni committed:
Apply hugovk suggestions from code review

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

antocuni committed:
Apply suggestions from code review

Co-authored-by: Adam Turner <9087854+AA-Turner@users.noreply.github.com>
Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

antocuni committed:
Apply suggestions from code review

Co-authored-by: Tomas R. <tomas.roun8@gmail.com>

antocuni committed:
remove unneeded lazy import

antocuni committed:
Update Lib/_pyrepl/fancycompleter.py

Co-authored-by: Hugo van Kemenade <1324225+hugovk@users.noreply.github.com>

antocuni committed:
move import to module scope

antocuni committed:
move import

antocuni committed:
kill this for now, we can redintroduce it later if/when we enable fancycompleter+pdb

antocuni committed:
this link is dead, add a comment to explain what it does instead

antocuni committed:
fix precommit

antocuni committed:
we need to make this import lazy, else we get circular imports

antocuni committed:
now that we have themes, we can kill the config object

antocuni committed:
style

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

antocuni committed:
fix mypy

antocuni committed:
document PYTHON_BASIC_COMPLETER

antocuni committed:
try to manually fix the filename

StanFromIreland commented on Doc/using/cmdline.rst:
```diff
@@ -1283,6 +1283,13 @@ conflict.
+.. envvar:: PYTHON_BASIC_COMPLETER
+
+   If this variable is set to any value, PyREPL will use :mod:`rlcompleter` to
+   implement tab complition, instead of the default :mod:`_pyrepl.fancycompleter`.
```
```suggestion
   implement tab complition, instead of the default.
```
I suggest not noting the the private module at all.
Otherwise, to fix the warning:
```suggestion
   implement tab complition, instead of the default :mod:`!_pyrepl.fancycompleter`.
```

antocuni replied:
thank you! Removed the link, and reworded the sentence slightly.

StanFromIreland commented on Doc/using/cmdline.rst:
```diff
@@ -1283,6 +1283,13 @@ conflict.
+.. envvar:: PYTHON_BASIC_COMPLETER
+
+   If this variable is set to any value, PyREPL will use :mod:`rlcompleter` to
+   implement tab complition, instead of the default :mod:`_pyrepl.fancycompleter`.
```
```suggestion
   implement tab completion, instead of the default :mod:`_pyrepl.fancycompleter`.
```
Typo

antocuni committed:
Typo

Co-authored-by: Stan Ulbrych <89152624+StanFromIreland@users.noreply.github.com>

antocuni committed:
reword

antocuni committed:
force PYTHON_COLORS=1 for tests which expects to see colors. Hopefully this fixes the failures on Android CI

antocuni committed:
fix it in a different way: just look in the theme to find the expected color. This is both more robust and more correct

antocuni committed:
fix precommit

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,174 @@
+            names += [' ']
+        return names
+
+    def colorize_matches(self, names, values):
```
How much time does this add to the autocompletion? I did a quick test with
```
>>> import sys
>>> sys.[TAB]
```
Then 86 names are processed with about 10 ms spend in `colorize_matches` alone. For modules or classes with more attributes the time would be longer. The processing does not seem needed since the response of the autocompletion is `[ not unique]`.

Are there any benchmarks for the full autocompletion or rules-of-thumb how long it should take?

(maybe these questions are better suited for a separate issue or followup PR)

pablogsal replied:
> Are there any benchmarks for the full autocompletion or rules-of-thumb how long it should take?

I don't think there are any but I think that as long as is not slow enough even if the module is too big I don't think it matters. We can always also deactivate if there are too many elements in the module

antocuni replied:
> Then 86 names are processed with about 10 ms spend in colorize_matches alone. 

this seems way too much.
I ran this super simple benchmarks on my machine:
```
import sys
import time
from _pyrepl.fancycompleter import Completer

def gendict(n):
    return {
        f'item{i}': i
        for i in range(n)
    }

class MyClass:
    pass

def main():
    obj = MyClass()
    comp_col = Completer({'obj': obj}, use_colors=True)
    comp_nocol = Completer({'obj': obj}, use_colors=False)

    for i in range(0, 10000, 1000):
        obj.__dict__ = gendict(i)
        a = time.perf_counter()
        comp_col.attr_matches('obj.')
        b = time.perf_counter()
        comp_nocol.attr_matches('obj.')
        c = time.perf_counter()

        t_col = b - a
        t_nocol = c - b
        diff_pct = (t_col - t_nocol) / t_nocol * 100
        print(f"{i:5d}: col={t_col*1000:.1f}ms  nocol={t_nocol*1000:.1f}ms  diff={diff_pct:+.2f}%")

main()
```

and I got this results:
```

❯ ./python bench_fancycompleter.py 
    0: col=0.1ms  nocol=0.1ms  diff=+18.30%
 1000: col=0.7ms  nocol=0.7ms  diff=+5.81%
 2000: col=1.6ms  nocol=1.4ms  diff=+13.75%
 3000: col=2.4ms  nocol=2.0ms  diff=+21.95%
 4000: col=2.9ms  nocol=2.4ms  diff=+19.63%
 5000: col=3.8ms  nocol=5.0ms  diff=-23.60%
 6000: col=5.3ms  nocol=4.5ms  diff=+15.84%
 7000: col=5.0ms  nocol=4.6ms  diff=+9.01%
 8000: col=6.9ms  nocol=7.6ms  diff=-9.75%
 9000: col=7.0ms  nocol=6.9ms  diff=+1.47%
```

so, on my machine it spends 7ms to process 9000 names, and the difference between colors and no colors is basically noise.

eendebakpt replied:
The example minimal example above never reaches `colorize_matches` because there is a common prefix.
If I replace `comp_col.attr_matches('obj.')` with `comp_col.attr_matches('obj.item')` I get this output
```
    0: col=0.1ms  nocol=0.0ms  diff=+95.88%
 2000: col=85.6ms  nocol=1.0ms  diff=+8489.94%
 4000: col=166.8ms  nocol=1.9ms  diff=+8763.17%
 6000: col=286.6ms  nocol=3.6ms  diff=+7940.23%
 8000: col=330.9ms  nocol=4.5ms  diff=+7297.60%
10000: col=429.9ms  nocol=5.1ms  diff=+8394.79%
12000: col=509.7ms  nocol=6.1ms  diff=+8308.30%
14000: col=632.8ms  nocol=9.4ms  diff=+6627.23%
16000: col=1308.4ms  nocol=29.5ms  diff=+4331.26%
18000: col=2131.1ms  nocol=35.8ms  diff=+5855.55%
```

antocuni replied:
oh right, sorry for the silly mistake. Indeed, this is what I get on my machine:
```
❯ ./python bench_fancycompleter.py 
    0: col=0.1ms  nocol=0.1ms  diff=+16.43%
 1000: col=9.6ms  nocol=1.0ms  diff=+826.21%
 2000: col=17.8ms  nocol=1.3ms  diff=+1293.67%
 3000: col=26.6ms  nocol=1.8ms  diff=+1377.77%
 4000: col=33.1ms  nocol=2.3ms  diff=+1335.15%
 5000: col=40.8ms  nocol=3.7ms  diff=+996.66%
 6000: col=54.1ms  nocol=3.7ms  diff=+1368.12%
 7000: col=56.0ms  nocol=4.1ms  diff=+1253.30%
 8000: col=63.4ms  nocol=6.7ms  diff=+851.07%
 9000: col=71.9ms  nocol=6.0ms  diff=+1095.03%
```

I'm not an UI expert at all, but I think that anything <100ms is perceived as "instantaneous" by the user, so unless we have modules with >9000 we should be safe 😅.

Jokes apart, I see that on your machine things are slower, but I still think that with "normal" number of completions the delay won't be noticeable (and, anecdotally, I never noticed in all these years).

eendebakpt replied:
Anecdotally, the timings are from my _fast_ machine, and I have noticed the new pyrepl being slower (not specifically this PR though). With the commit moving the `gettheme()` out of the loop performance is much better:
```
    0: col=0.1ms  nocol=0.0ms  diff=+114.87%
 1000: col=1.8ms  nocol=0.5ms  diff=+257.26%
 2000: col=3.2ms  nocol=1.1ms  diff=+187.30%
 3000: col=3.9ms  nocol=1.4ms  diff=+182.85%
 4000: col=4.9ms  nocol=2.0ms  diff=+143.63%
 5000: col=6.8ms  nocol=2.9ms  diff=+137.13%
 6000: col=8.0ms  nocol=3.2ms  diff=+152.49%
 7000: col=10.0ms  nocol=3.5ms  diff=+185.24%
 8000: col=10.9ms  nocol=4.3ms  diff=+153.73%
 9000: col=12.1ms  nocol=4.8ms  diff=+149.98%
```
So I am happy with the current performance of the PR!

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,174 @@
+                   in enumerate(zip(names, values))]
+        # We add a space at the end to prevent the automatic completion of the
+        # common prefix, which is the ANSI escape sequence.
+        return matches + [' ']
```
```suggestion
        matches.append(' ')
        return matches
```
This avoids creation of the list, is a bit faster.

antocuni committed:
Update Lib/_pyrepl/fancycompleter.py

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

hugovk commented on Lib/_colorize.py:
```diff
@@ -219,6 +219,30 @@ class Unittest(ThemeSection):
 
 
+@dataclass(frozen=True, kw_only=True)
+class FancyCompleter(ThemeSection):
```
Please keep this class in alphabetical order, like the others.

Same for the `fancycompleter` below.

antocuni replied:
done in 3a6bcd324bb

antocuni committed:
put _colorize.FancyCompleter in alphabetical order w.r.t. the other sections

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,175 @@
+            return self.colorize_matches(names, values)
+
+        if prefix:
+            names += [' ']
```
```suggestion
            names.append(' ')
```

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,175 @@
+        return f"{N}{color}{name}{ANSIColors.RESET}"
+
+    def color_by_type(self, t):
+        theme = get_theme()
```
Getting the theme is quite expensive. Creating the theme in `colorize_matches` and passing it to `color_for_obj` and `color_by_type` greatly improves performance (for a large number of matches).

Also consider making `color_by_type` (and others) private methods, since there are not used outside this class.

antocuni replied:
fixed in 743e661dc29

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,175 @@
+        return getattr(theme.fancycompleter, typename, ANSIColors.RESET)
+
+
+def commonprefix(names, base=''):
```
The argument `base` is not used in the code. Can we remove it?

antocuni replied:
fixed in 2ebf50e3bd4

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,175 @@
+            noprefix = '__'
+        else:
+            noprefix = None
+        words = sorted(words)
```
The sorting is not needed in all cases (e.g. when names ends up with only 1 entry, or there is a common prefix), so we can sort the names when returning instead of sorting the `words` here.

(`words` is a set as input, but we could rewrite the above code to avoid creating a set)

antocuni replied:
this starts to feel a bit nitpicking, but fixed in 6a5bcfe9ed2

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,175 @@
+        if '(' in expr or ')' in expr:  # don't call functions
+            return []
+        try:
+            thisobject = eval(expr, self.namespace)
```
Could this be the following?
```suggestion
            thisobject = self.namespace.get(expr)
```

antocuni replied:
no, it doesn't work in case of e.g. `a.b.c.<TAB>`

eendebakpt replied:
I see. In general I like to avoid using `eval` in code (because of safety and performance reasons). Not using eval here would add a few lines of code though. Leaving the decision on this to a core dev

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,175 @@
+            return []
+
+        if len(names) == 1:
+            return [f'{expr}.{names[0]}']  # only option, no coloring.
```
Should we color anyway for consistency?

antocuni replied:
no. When you return a single completion, readline inserts it in the prompt. If we return colors here, we would be ANSI codes in the prompt, which is surely not what you want

eendebakpt replied:
Agreed, changing the colors in the prompt is not what we want. Could you update the comment with the reason?

antocuni committed:
get_theme() is relatively expensive, fetch it early and cache it

antocuni committed:
base is never used when calling commonprefix, remove it

antocuni committed:
Update Lib/_pyrepl/fancycompleter.py

Co-authored-by: Pieter Eendebak <pieter.eendebak@gmail.com>

antocuni committed:
there is no need to sort words in advance, we can just sort names later

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -131,6 +130,7 @@ def attr_matches(self, text):
         if prefix and prefix != attr:
             return [f'{expr}.{prefix}']  # autocomplete prefix
 
+        names.sort()
```
This does not sort the `values`. Maybe revert the commit that introduced this if it is too much trouble to fix this.

antocuni replied:
yes, pro tip for my future self: don't write code when you are sleepy :).
commit 7d2c7902817 reverts it, and also adds a comment to explain WHY we need to sort words.

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,176 @@
+        else:
+            return rlcompleter.Completer.complete(self, text, state)
+
+    def _callable_postfix(self, val, word):
+        # disable automatic insertion of '(' for global callables
+        return word
+
```
```suggestion
```
The method does not seem to be used.

antocuni replied:
yes, it's used. We are subclassing `rlcompleter.Completer` and by overriding this method we prevent the insertion of '('.

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,176 @@
+        values = []
+        for name in names:
+            clean_name = name.rstrip(': ')
+            if clean_name in keyword.kwlist:
```
```suggestion
            if keyword.iskeyword(clean_name):
```

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,176 @@
+        words = set(dir(thisobject)) - {'__builtins__'}
+
+        if hasattr(thisobject, '__class__'):
+            words.add('__class__')
```
If I remove this line, tests still pass. Why is this needed?

antocuni replied:
this is a valid question, and I admit I don't know the answer.
The code was originally introduced by this PR in fancycompleter, which merged some logic from CPython's `rlcompleter.py`:
https://github.com/pdbpp/fancycompleter/pull/17

The current rlcompleter.py contains exactly the same lines:
https://github.com/python/cpython/blob/7d2c790281797769b6cfa5e76f3c16aad2c17ff6/Lib/rlcompleter.py#L162-L167

So, I think it's a good idea to keep them in sync.

eendebakpt replied:
The behaviour in the rlcompleter was added 25 years ago in https://github.com/python/cpython/commit/4e20de59dd5a9058c0d99c10d773b9d0bcbbe07d and https://github.com/python/cpython/commit/46bd9a6191236c07401169ff407914aeacbddc98 (for the `__builtins__`). I will check later whether the behaviour still makes sense, but changing that would be a different PR.

Can you use the same style as in rlcompleter.py though? E.g. use `words.discard`.

antocuni replied:
It *was* using `words.discard`, but then I committed this suggestion by @AA-Turner 
https://github.com/python/cpython/pull/130473#discussion_r1972577973

I'm happy with both :man_shrugging:

AA-Turner replied:
If we can remove the line in `fancycompleter` and nothing breaks, we may as well do so.

eendebakpt replied:
Both the ` words.add('__class__')` and `words.discard("__builtins__")` can be removed in `rlcompleter.py` without tests breaking (locally for me). The `words.add('__class__') ` seems rather safe to me (if `__class__` is an attribute of `thisobject`, it should be in `dir(thisobject)`).

Removing `words.discard("__builtins__")` is also fine with me, but I have no knowledge of the history here. The commit states

> Do not expose __builtins__ name as a completion; this is an implementation detail that confuses too many people.  Based on discussion in python-dev.

The commit is so old, I could not find the python-dev archives containing the rationale. On modern Python there are a few more dunder names the might be confusing (e.g. in the global namespace we have `__spec__`, `__package__`, `__build_class__` and a few more), so why would `__builtins__` be special. Maybe in the python version back then `__builtins__` was more prominent? Again, it is so long ago I cannot install a python version from that time on my system.

eendebakpt replied:
Note: if we decide to remove these, I would do it in a separate PR.

eendebakpt commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,176 @@
+        except Exception:
+            return []
+
+        # get the content of the object, except __builtins__
```
Why do we have to remove `__builtins__`? If I remove this part the unit tests still pass.

antocuni replied:
same answer as above, it's using the same logic as rlcompleter

antocuni committed:
undo 6a5bcfe9ed: there IS actually a good reason to sort the words in advance, let's add a comment to explain why

pablogsal committed:
gh-130472: Fix pyrepl fancycompleter edge cases

Keep pyrepl completion logic working on the uncolored completion text.
The reader now strips ANSI escapes before comparing the typed stem,
inserting a sole completion, computing the shared prefix, and filtering
an open completion menu. This fixes colored completions that would stop
refining correctly once more characters were typed.

Restore readline's callable postfix behavior for attribute completions by
routing single attribute matches through rlcompleter's callable postfix
logic while keeping the full expr.attr stem for menu refinement. Global
completion also treats soft keywords as keywords instead of trying to
evaluate them.

Avoid side effects while probing attribute values for coloring by not
forcing property access and by preserving lazy module imports. Also make
the fake color-sorting escape prefix round-trip cleanly once the match
index grows past three digits.

Only honor PYTHON_BASIC_COMPLETER when the environment is enabled, so
pyrepl setup now respects -E / sys.flags.ignore_environment. Add
regression tests for the reader behavior, callable attribute completion,
property and lazy-import safety, large color-sort prefixes, and the -E
setup path.

pablogsal reviewed (approved):
LGTM after my fixes

StanFromIreland commented on Lib/_pyrepl/fancycompleter.py:
```diff
@@ -0,0 +1,210 @@
+    def _color_for_obj(self, i, name, value):
+        t = type(value)
+        color = self._color_by_type(t)
+        # Encode the match index into a fake escape sequence that
+        # stripcolor() can still remove once i reaches four digits.
+        N = f"\x1b[{i // 100:03d};{i % 100:02d}m"
```
These are in fact not always "fake," leading to the issues in my comment above. Fixing that (ensuring they are greater than 100, which IIRC is the limit of valid sequences), I still have contrast issues, so maybe there is more to it:

<img width="1322" height="421" alt="Image" src="https://github.com/user-attachments/assets/126aba92-8d6a-4219-a11c-812526f5d540" />

pablogsal replied:
Let's open an issue and fix this in a separate PR
